cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -2709,34 +2760,36 @@ private List<Field<?>> getRunRecordApplicationRefPrefix(String status,
   }
 
   private List<Field<?>> getRunRecordProgramPrefix(String status, @Nullable ProgramId programId) {
-    List<Field<?>> fields = getRunRecordStatusPrefix(status);
-    if (programId == null) {
-      return fields;
-    }
-    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.NAMESPACE_FIELD,
-        programId.getNamespace()));
-    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.APPLICATION_FIELD,
-        programId.getApplication()));
-    fields.add(
-        Fields.stringField(StoreDefinition.AppMetadataStore.VERSION_FIELD, programId.getVersion()));
-    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.PROGRAM_TYPE_FIELD,
-        programId.getType().name()));
-    fields.add(
-        Fields.stringField(StoreDefinition.AppMetadataStore.PROGRAM_FIELD, programId.getProgram()));
-    return fields;
+    ProgramReference programRef = programId != null ? programId.getProgramReference() : null;
+    String version = programId != null ? programId.getVersion() : null;
+
+    return getRunRecordProgramPrefixInternal(status, programRef, version);
   }
 
   private List<Field<?>> getRunRecordProgramRefPrefix(String status,
       ProgramReference programReference) {
+    return  getRunRecordProgramPrefixInternal(status, programReference, null);
+  }
+
+  private List<Field<?>> getRunRecordProgramPrefixInternal(String status,
+      @Nullable ProgramReference programRef, @Nullable String version) {
```

Comment:
I don't think you need "internal". All methods are private, so one is no more internal than other. 
Also given amount of overloading, mixing it up with nullability is confusing.  Can `getRunRecordProgramPrefix(String status, @Nullable ProgramId programId)` just call getRunRecordStatusPrefix directly? This way this method can have non-null parameters.
Overall, I'd follow the pattern of non-nullable parameters and then caller add whatever it needs (e.g. version)

Reply:
Done